### ISSUE TITLE
Logging does not preserve escaped quotes

### ISSUE THREAD

**Issue by gamesh411** [Ericsson/codechecker#2025]
The generation of asts for the following build action fails due to incorrectly defined macro value.

The make invocation:
```
gcc -DPACKAGE_NAME=\"tmux\" -DPACKAGE_TARNAME=\"tmux\" -DPACKAGE_VERSION=\"2.6\" -DPACKAGE_STRING=\"tmux\ 2.6\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DPACKAGE=\"tmux\" -DVERSION=\"2.6\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DIRENT_H=1 -DHAVE_FCNTL_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_PATHS_H=1 -DHAVE_PTY_H=1 -DHAVE_STDINT_H=1 -DHAVE_SYS_DIR_H=1 -DHAVE_DIRFD=1 -DHAVE_FLOCK=1 -DHAVE_PRCTL=1 -DHAVE_SYSCONF=1 -DHAVE_ASPRINTF=1 -DHAVE_CFMAKERAW=1 -DHAVE_EXPLICIT_BZERO=1 -DHAVE_MEMMEM=1 -DHAVE_REALLOCARRAY=1 -DHAVE_SETENV=1 -DHAVE_STRCASESTR=1 -DHAVE_STRNDUP=1 -DHAVE_STRSEP=1 -DHAVE_NCURSES_H=1 -DHAVE_B64_NTOP=1 -DHAVE_DAEMON=1 -DHAVE_FORKPTY=1 -DHAVE___PROGNAME=1 -DHAVE_PROGRAM_INVOCATION_SHORT_NAME=1 -DHAVE_PR_SET_NAME=1 -DHAVE_PROC_PID=1 -I.  -DTMUX_CONF="\"/etc/tmux.conf\""  -iquote.      -D_GNU_SOURCE -D_DEFAULT_SOURCE   -std=gnu99 -O2    -MT cfg.o -MD -MP -MF $depbase.Tpo -c -o cfg.o cfg.c
```

The relevant line inside the logged compile_commands.json:
```
{
		"directory": "/home/efulop/projects/tmux",
		"command": "/usr/bin/ccache -DPACKAGE_NAME=\"tmux\" -DPACKAGE_TARNAME=\"tmux\" -DPACKAGE_VERSION=\"2.6\" -DPACKAGE_STRING=\"tmux\\ 2.6\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DPACKAGE=\"tmux\" -DVERSION=\"2.6\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DIRENT_H=1 -DHAVE_FCNTL_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_PATHS_H=1 -DHAVE_PTY_H=1 -DHAVE_STDINT_H=1 -DHAVE_SYS_DIR_H=1 -DHAVE_DIRFD=1 -DHAVE_FLOCK=1 -DHAVE_PRCTL=1 -DHAVE_SYSCONF=1 -DHAVE_ASPRINTF=1 -DHAVE_CFMAKERAW=1 -DHAVE_EXPLICIT_BZERO=1 -DHAVE_MEMMEM=1 -DHAVE_REALLOCARRAY=1 -DHAVE_SETENV=1 -DHAVE_STRCASESTR=1 -DHAVE_STRNDUP=1 -DHAVE_STRSEP=1 -DHAVE_NCURSES_H=1 -DHAVE_B64_NTOP=1 -DHAVE_DAEMON=1 -DHAVE_FORKPTY=1 -DHAVE___PROGNAME=1 -DHAVE_PROGRAM_INVOCATION_SHORT_NAME=1 -DHAVE_PR_SET_NAME=1 -DHAVE_PROC_PID=1 -I/home/efulop/projects/tmux -DTMUX_CONF=\"/etc/tmux.conf\" -iquote. -D_GNU_SOURCE -D_DEFAULT_SOURCE -std=gnu99 -O2 -MT cfg.o -MD -MP -MF .deps/cfg.Tpo -c -o /home/efulop/projects/tmux/cfg.o /home/efulop/projects/tmux/cfg.c",
		"file": "/home/efulop/projects/tmux/cfg.c"
	}
```

The action parameter printed out inside ctu_manager.py in map_functions:
```
Original command: /usr/bin/ccache -DPACKAGE_NAME="tmux" -DPACKAGE_TARNAME="tmux" -DPACKAGE_VERSION="2.6" -DPACKAGE_STRING="tmux\ 2.6" -DPACKAGE_BUGREPORT="" -DPACKAGE_URL="" -DPACKAGE="tmux" -DVERSION="2.6" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DIRENT_H=1 -DHAVE_FCNTL_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_PATHS_H=1 -DHAVE_PTY_H=1 -DHAVE_STDINT_H=1 -DHAVE_SYS_DIR_H=1 -DHAVE_DIRFD=1 -DHAVE_FLOCK=1 -DHAVE_PRCTL=1 -DHAVE_SYSCONF=1 -DHAVE_ASPRINTF=1 -DHAVE_CFMAKERAW=1 -DHAVE_EXPLICIT_BZERO=1 -DHAVE_MEMMEM=1 -DHAVE_REALLOCARRAY=1 -DHAVE_SETENV=1 -DHAVE_STRCASESTR=1 -DHAVE_STRNDUP=1 -DHAVE_STRSEP=1 -DHAVE_NCURSES_H=1 -DHAVE_B64_NTOP=1 -DHAVE_DAEMON=1 -DHAVE_FORKPTY=1 -DHAVE___PROGNAME=1 -DHAVE_PROGRAM_INVOCATION_SHORT_NAME=1 -DHAVE_PR_SET_NAME=1 -DHAVE_PROC_PID=1 -I/home/efulop/projects/tmux -DTMUX_CONF="/etc/tmux.conf" -iquote. -D_GNU_SOURCE -D_DEFAULT_SOURCE -std=gnu99 -O2 -MT cfg.o -MD -MP -MF .deps/cfg.Tpo -c -o /home/efulop/projects/tmux/cfg.o /home/efulop/projects/tmux/cfg.c,
Analyzer type: clangsa,
 Analyzer options: ['-DPACKAGE_NAME=tmux', '-DPACKAGE_TARNAME=tmux', '-DPACKAGE_VERSION=2.6', '-DPACKAGE_STRING=tmux\\ 2.6', '-DPACKAGE_BUGREPORT=', '-DPACKAGE_URL=', '-DPACKAGE=tmux', '-DVERSION=2.6', '-DSTDC_HEADERS=1', '-DHAVE_SYS_TYPES_H=1', '-DHAVE_SYS_STAT_H=1', '-DHAVE_STDLIB_H=1', '-DHAVE_STRING_H=1', '-DHAVE_MEMORY_H=1', '-DHAVE_STRINGS_H=1', '-DHAVE_INTTYPES_H=1', '-DHAVE_STDINT_H=1', '-DHAVE_UNISTD_H=1', '-D__EXTENSIONS__=1', '-D_ALL_SOURCE=1', '-D_GNU_SOURCE=1', '-D_POSIX_PTHREAD_SEMANTICS=1', '-D_TANDEM_SOURCE=1', '-DHAVE_DIRENT_H=1', '-DHAVE_FCNTL_H=1', '-DHAVE_INTTYPES_H=1', '-DHAVE_PATHS_H=1', '-DHAVE_PTY_H=1', '-DHAVE_STDINT_H=1', '-DHAVE_SYS_DIR_H=1', '-DHAVE_DIRFD=1', '-DHAVE_FLOCK=1', '-DHAVE_PRCTL=1', '-DHAVE_SYSCONF=1', '-DHAVE_ASPRINTF=1', '-DHAVE_CFMAKERAW=1', '-DHAVE_EXPLICIT_BZERO=1', '-DHAVE_MEMMEM=1', '-DHAVE_REALLOCARRAY=1', '-DHAVE_SETENV=1', '-DHAVE_STRCASESTR=1', '-DHAVE_STRNDUP=1', '-DHAVE_STRSEP=1', '-DHAVE_NCURSES_H=1', '-DHAVE_B64_NTOP=1', '-DHAVE_DAEMON=1', '-DHAVE_FORKPTY=1', '-DHAVE___PROGNAME=1', '-DHAVE_PROGRAM_INVOCATION_SHORT_NAME=1', '-DHAVE_PR_SET_NAME=1', '-DHAVE_PROC_PID=1', '-I/home/efulop/projects/tmux', '-DTMUX_CONF=/etc/tmux.conf', '-iquote.', '-D_GNU_SOURCE', '-D_DEFAULT_SOURCE', '-std=gnu99', '-O2']
```

The command that generates the ast:
```
clang -nobuiltininc -isystem /home/efulop/projects/upstream-clang-build-release/lib/clang/9.0.0/include -c -x c -DPACKAGE_NAME=tmux -DPACKAGE_TARNAME=tmux -DPACKAGE_VERSION=2.6 -DPACKAGE_STRING=tmux\ 2.6 -DPACKAGE_BUGREPORT= -DPACKAGE_URL= -DPACKAGE=tmux -DVERSION=2.6 -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DIRENT_H=1 -DHAVE_FCNTL_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_PATHS_H=1 -DHAVE_PTY_H=1 -DHAVE_STDINT_H=1 -DHAVE_SYS_DIR_H=1 -DHAVE_DIRFD=1 -DHAVE_FLOCK=1 -DHAVE_PRCTL=1 -DHAVE_SYSCONF=1 -DHAVE_ASPRINTF=1 -DHAVE_CFMAKERAW=1 -DHAVE_EXPLICIT_BZERO=1 -DHAVE_MEMMEM=1 -DHAVE_REALLOCARRAY=1 -DHAVE_SETENV=1 -DHAVE_STRCASESTR=1 -DHAVE_STRNDUP=1 -DHAVE_STRSEP=1 -DHAVE_NCURSES_H=1 -DHAVE_B64_NTOP=1 -DHAVE_DAEMON=1 -DHAVE_FORKPTY=1 -DHAVE___PROGNAME=1 -DHAVE_PROGRAM_INVOCATION_SHORT_NAME=1 -DHAVE_PR_SET_NAME=1 -DHAVE_PROC_PID=1 -I/home/efulop/projects/tmux -DTMUX_CONF=/etc/tmux.conf -iquote. -D_GNU_SOURCE -D_DEFAULT_SOURCE -std=gnu99 -O2 /home/efulop/projects/tmux/cfg.c -emit-ast -D__clang_analyzer__ -w -o /home/efulop/projects/tmux/results/ctu-dir/x86_64/ast/home/efulop/projects/tmux/cfg.c.ast
```

The defined macro value here is expected to represent a cstring, so when expanded the enclosing quotes should be there. This is the case with the original make command, but it gets lost somehow. I suspect, that inside the logged compile_commands.json the original form should remain (like this:   `-DTMUX_CONF="\"...\""`    compared to:    `-DTMUX_CONF=\"...\"`     ).

**Comment by bruntib** [comment]
This is fixed by #2027. Could you please check the master version? If it works then this ticket can be closed.

**Comment by gamesh411** [comment]
Fixed the issue. Thanks for the quick response 👍.
